Feat: New GM Commands, princess commands on map menu, graphics for some explosions

Scoppio opened the pull request:
# New GM Commands & Much More!
All the commands listed here are GM only, they allow the player who is a Game Master more freedom to change the game to the way they want the story to play out. 
They would allow co-op games and RPG or dramatized games to have more events and give the GM more control over how they go.

Examples:
- Custom scenario, a battle is happening while a space battle goes, every turn there is a chance a stray shot from a warship hits the ground mid battle.

- A F5 hurrinane is moving towards the base and you need to evacuate the vehicles and people as fast as possible before they all die.

- In the middle of the battle the day becomes night as the moon engulfs the sun.

- The star of the system is pulsating, during round 10 it goes super nova, the planet you are playing is far away enough to not be instantaneously destroyed, but the temperature gets extreme and the atmosphere of the planetoid is stripped instantaneously, everyone that was outside a battlemek is dead. Is it worth keeping the fight?

## GamemasterServerCommand
New type of ServerCommand, can only be used by gamemaster, it also has an automatic parser for its arguments with automatic error handling. Also has automated "documentation", using the descriptions for each variable as input for itself.


<img width="763" alt="Screenshot 2024-11-14 at 23 49 54" src="https://github.com/user-attachments/assets/9a606e23-45c1-4ff7-a257-69042f97fd54">

<img width="715" alt="Screenshot 2024-11-14 at 23 50 39" src="https://github.com/user-attachments/assets/c6d4e709-8b76-4c55-99a0-8b995a1ef878">

This extra complexity also allows for an automatic "dialog" setup that I implemented (its only in a functional state, not a good state).
<img width="553" alt="Screenshot 2024-11-14 at 23 51 13" src="https://github.com/user-attachments/assets/444fca00-2f36-4633-a444-8ff8f0c1315c">

- Added a new type of server command, the gamemaster server command, it has automated parsing, error handling, argument assertion, automated documentation (for \/help), it also implements the necessary tools to automate the GM Menu commands.

- New server commands accept both positional arguments and named arguments. Positional and named arguments can be used in the same command but positional arguments must come first. Named arguments may be used in any order (after positional arguments if those are being used). Example:

```
The following are equivalent and valid:

    /ob 5 10 200 5
    /ob x=5 y=10 dmg=200 radius=5
    /ob 5 10 dmg=200 radius=5
    /ob 5 dmg=200 radius=5 y=10
    /ob 5 10 radius=5 dmg=200

```


## Command KILL - When you want something DEAD
**\/kill \<id\>**
Kills (ID) unit at GM's discretion, needs to be enabled in the game options.

## Command NOSMOKE - Clear skies
**\/nosmoke**
Clears all smoke clouds in the board. 

## Command WEATHER - Change planetary conditions
**\/weather [...args]**
Allows the GM to change the weather on demand

## Command Disaster - Disasters
**\/gomorrah \<type\>**
Cast a disaster (like in the old days of SimCity) to spicy things up (WIP)!
The cheat is a homage to SimCity "gomorrah" cheat code which would start a nuclear disaster.

## Command OB - Orbital Bombardment
**\/ob \<x\> \<y\> [damage=#] [radius=4]**
Calls in orbital bombardment at the end of the next weapons firing phase (WIP)!

## Command ChangeOwner - Traitor 2.0
**\/changeOwner \<UnitID\> \<PlayerID\>**
Gives a unit to a player.


## Command Rescue
**\/rescue \<UnitID\>**
Rescue from the board a single unit. A "forced flee" if you will.


---
# (Very crude) GM command dialogs
Allows the gamemaster to manually enter the commands

Features:
- Numerical spinners which obeys the limits set in the command arguments.
- Enumerator picker/drop down for those more complex commands.
- Numerical spinners for X and Y position defaults to current hex square (under the mouse when you right clicked on the map).


<img width="411" alt="Screenshot 2024-11-14 at 23 26 11" src="https://github.com/user-attachments/assets/a0b476c4-8077-4e5e-bdf0-24ee518baf39">


## GM Contextual Menu (Map Menu)

### Ps: Some screenshots were taken early in development, so mileage may vary.

Added more functions to the map menu for GMs, most notable being:
- Removed "Traitor" entry from Map Menu. I can roll back this change, but it was made because the current traitor functionality is very limited and the way it currently works does allow for big mistakes for those not used to how it works.
- Added a new "Traitor" inside the GM menu, this uses a more contextual approach, it allows you to pick a unit in the hex as you would when targeting a unit.

<img width="761" alt="Screenshot 2024-11-14 at 23 48 00" src="https://github.com/user-attachments/assets/5247c829-b08f-4bcc-b8ae-88c001a312b3">

- Added new "Kill Unit" inside the GM menu. Allows you to pick a unit in the hex as you would when targeting a unit.
<img width="689" alt="Screenshot 2024-11-14 at 23 47 17" src="https://github.com/user-attachments/assets/3d124fa2-f8b8-4d98-a57b-c932958be0d3">

- Added new "Special Commands" menu to the GM Menu. It lists all commands that are implemented with GamemasterServerCommand class.

<img width="473" alt="Screenshot 2024-11-14 at 23 46 29" src="https://github.com/user-attachments/assets/fbc11bc5-868e-488c-9cda-062daf2e5f46">

<img width="639" alt="Screenshot 2024-11-14 at 23 49 18" src="https://github.com/user-attachments/assets/8b34a1d9-dfec-4d6f-a22c-273c7cce8c22">

# Quick Access to Princess controls 

- You can now give orders to your allied royalty (Bots) using the right click menu, making the process of changing Princess' behavior, priority targets, etc, way easier.

<img width="918" alt="Screenshot 2024-11-17 at 02 03 51" src="https://github.com/user-attachments/assets/5b2951e8-fdf8-47d2-941d-0ec86e6a3ad3">

# Graphics for BIG BOOM

Nukes and orbital bombardment have graphics.

![image](https://github.com/user-attachments/assets/c14383ac-e941-4856-9e1d-e55076fba47f)
![image](https://github.com/user-attachments/assets/bd82713d-e27d-4ce6-b53e-e1177ea99b7f)

# Support for BIG BOOM

Exploding a nuke on the game is a sure way to grind everything to a halt, your big powerful PC may simply sit down crunching numbers with CPU at 100% for many minutes, and the result is a 5 miles long report with all the units taking tens of thousands of damage points. I added an early exit check that identifies if the unit is already "devastated", if true then it consider the damage processing for that unit done and goes for the next one.
With this small change, now releasing nukes that boil lakes and turn cities into craters is nothing but childsplay!

IllianiCBT commented on megamek/src/megamek/server/commands/ChangeWeatherCommand.java:
```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (C) 2024 Luana Scoppio (luana.coppio@gmail.com)
```
Are you claiming copyright of this code? Normally we ask that the code be under the MegaMek team copyright, I believe

IllianiCBT replied:
Just confirmed with the team, this isn’t the issue I thought it was, so carry on! :)

Scoppio replied:
I kinda copied and pasted what were in other files, since this is based directly in an older file that is the same license I tought that adding my name to it to make clear who authored would be reasonable.
But either way, I will review if I didnt forget the megamek team entry in any of the files.

SJuliez commented on megamek/i18n/megamek/common/options/messages.properties:
```diff
@@ -140,6 +140,8 @@ GameOptionsInfo.option.allow_nukes.displayableName=Allow command-line nukes.
 GameOptionsInfo.option.allow_nukes.description=This must be checked to allow players to throw nukes from the command line. If this is not checked, nukes are still available as advanced munitions.
 GameOptionsInfo.option.really_allow_nukes.displayableName=REALLY allow command-line nukes.
 GameOptionsInfo.option.really_allow_nukes.description=This must be checked to allow players to throw nukes from the command line. If this is not checked, nukes are still available as advanced munitions.
+GameOptionsInfo.option.gm_can_kill_units.displayableName=Allow command-line kill (GM only).
```
I suggest lumping this together with command line nukes. I doubt the distinction is required.

gsparks3 replied:
Personally I would go with "GM gets access to all GM commands by default". Nominally, the point is to give them access to various tools that let them arbitrarily alter the scenario as needed. The GM status itself already has some safeguards applied to prevent abuse (can't just activate it midgame without the players voting to allow it, etc.).

IllianiCBT reviewed (commented):
I saw you requested I review this.

I don’t have enough working knowledge of the mm side of our codebase, so try not to review mm-specific stuff unless it’s super basic - and even then, I prefer not to get involved. Didn’t want you thinking I was just ignoring your request. :)

BLR-IIC commented:
I'm pretty excited to see this in progress.  I've been turning over code in my head for making command line orbital bombardment work.  I was going to borrow code from the nuke commands.  I hope these commands see implementation!

HammerGS commented:
Random thought...how hard would it be to create a "GM Command GUI" a simple GUI that lists the commands and lets you input the values

Scoppio commented:
> Random thought...how hard would it be to create a "GM Command GUI" a simple GUI that lists the commands and lets you input the values

I think its doable, I will do those in small steps. I want to release those first the way they are. I just need to make sure the terrain damage is not too bad.

Scoppio commented:
> I'm pretty excited to see this in progress. I've been turning over code in my head for making command line orbital bombardment work. I was going to borrow code from the nuke commands. I hope these commands see implementation!

Thanks, I picked the nuke as an example and turns out it is pretty standard. Actually the nuke doesnt take things into considerations like VTOL. It also is very expensive because it reachs the entire board and the scale of it is madness (amount of damage, garanteed kill on hit, etc).

So I had to reimplement alot of stuff, but there is still too much code duplication here, and not everything of the Terrain Factor and terrain damage is properly implemented.

I think I will implement the damage terrain in small steps until it is completely implemented, but at least it will be "good enough" for the narrative play.

Scoppio commented on megamek/src/megamek/client/ui/swing/MapMenu.java:
```diff
@@ -433,21 +410,49 @@
         }
     }
 
+    private record Tuple(String name, GamemasterServerCommand command) {}
```
I need to remove this

Scoppio commented on megamek/src/megamek/client/ui/swing/boardview/BoardView.java:
```diff
@@ -1454,6 +1456,38 @@
+    private void drawOrbitalBombardmentHexes(Graphics boardGraphics) {
+        Image orbitalBombardmentImage = tileManager.getOrbitalBombardmentImage();
+        Rectangle view = boardGraphics.getClipBounds();
+        boolean justDraw = false;
```
Also need to delete this.

gsparks3 commented:
Saw you requested a review from me - while I know how to code, I don't really know Java at all, so I can't comment on the code itself. However, even as a draft, the features being added look very promising. The right-click context menus are an excellent way to implement the location-specific GM commands; specifying hex coordinates manually is always clunky. I know my GM will be very happy to have the "kill" command to remove unwanted/unneeded units (e.g. a retreating foot platoon that isn't being attacked and would take 20 turns to reach the map edge) and especially the orbital bombardment command.

Scoppio commented on megamek/src/megamek/client/ui/swing/boardview/BoardView.java:
```diff
@@ -1454,6 +1456,44 @@ private Mounted<?> selectedWeapon() {
         return (clientgui != null) ? clientgui.getDisplayedWeapon().orElse(null) : null;
     }
 
+    /**
+     * Draw the orbital bombardment attacks on the board view
+     *
+     * @author Luana Coppio
+     * @param boardGraphics     The graphics object to draw on
+     */
+    private void drawOrbitalBombardmentHexes(Graphics boardGraphics) {
```
I dont think this is working...

Scoppio commented on megamek/src/megamek/client/ui/swing/minimap/Minimap.java:
```diff
@@ -536,6 +536,9 @@ private void drawMap(boolean forceDraw) {
             for (int j = 0; j < board.getWidth(); j++) {
                 for (int k = 0; k < board.getHeight(); k++) {
                     Hex h = board.getHex(j, k);
+                    if (h == null) {
+                        continue;
+                    }
```
Missing nullpointer check. It should never be null, but had it come back null to me a couple of times...

Scoppio commented:
> Saw you requested a review from me - while I know how to code, I don't really know Java at all, so I can't comment on the code itself. However, even as a draft, the features being added look very promising. The right-click context menus are an excellent way to implement the location-specific GM commands; specifying hex coordinates manually is always clunky. I know my GM will be very happy to have the "kill" command to remove unwanted/unneeded units (e.g. a retreating foot platoon that isn't being attacked and would take 20 turns to reach the map edge) and especially the orbital bombardment command.

I believe I requested your review because the way you left a comment Github listed you as a reviewer (so after I posted a few extra things I just re-requested reviews). Sorry for the notification.

Scoppio commented:
> Saw you requested a review from me - while I know how to code, I don't really know Java at all, so I can't comment on the code itself. However, even as a draft, the features being added look very promising. The right-click context menus are an excellent way to implement the location-specific GM commands; specifying hex coordinates manually is always clunky. I know my GM will be very happy to have the "kill" command to remove unwanted/unneeded units (e.g. a retreating foot platoon that isn't being attacked and would take 20 turns to reach the map edge) and especially the orbital bombardment command.

I am very sorry, I have no idea how I requested your review

HammerGS commented:
This was a request from the forums 

If reworking the GM commands, can we get the Princess Ignore command on a right click menu somehow?

Scoppio commented:
> This was a request from the forums
> 
> If reworking the GM commands, can we get the Princess Ignore command on a right click menu somehow?

I will take a look, depending of complexity I can push for a next release

Scoppio commented:
Added Nuke graphics, added controls for allied Princesses on the contextual map menu (right mouse button click on the map).

Scoppio commented on megamek/src/megamek/client/ui/swing/CollapseWarning.java:
```diff
@@ -96,7 +96,7 @@ public static List<Coords> findCFWarningsMovement(Game g, Entity e, Board b) {
             List<Coords> hexesToCheck = new ArrayList<Coords>();
             if (pos != null) {
-                hexesToCheck = pos.allAtDistanceOrLess(range + 1);
+                hexesToCheck = pos.allAtDistanceOrLess(range);
```
Fixed the bug in the function so the fix isnt necessary anymore

Scoppio commented on megamek/src/megamek/client/ui/swing/MapMenu.java:
```diff
@@ -420,32 +408,287 @@ private JMenu createSpecialHexDisplayMenu() {
-            return menu;
-        } else {
-
+        if (client.getLocalPlayer().getGameMaster()) {
```
I need to move those things to another class, will do it on a next PR

Scoppio commented on megamek/src/megamek/client/ui/swing/boardview/TerrainShadowHelper.java:
```diff
@@ -185,9 +185,11 @@ BufferedImage updateShadowMap() {
-                    int lv = nhex.getLevel();
-                    if (lv < level) {
-                        surrounded = false;
+                    if (nhex != null) {
```
another instance where it may very well be null and there was no null guard

Scoppio commented on megamek/src/megamek/common/Coords.java:
```diff
@@ -485,44 +486,51 @@ public boolean isOnHexRow(int direction, @Nullable Coords other) {
+     * Returns a list of all coordinates at the given distance (dist - 1)
+     * and anything less than dist as well.
+     */
+    public ArrayList<Coords> allLessThanDistance(int dist) {
```
places in the code that were using the previous version of the function `allAtDistanceOrLess` were changed to use this one instead. This implements the behavior that the previous/bugged version had (same functionality)

Scoppio commented on megamek/src/megamek/common/Mek.java:
```diff
@@ -4185,6 +4185,9 @@ public boolean canAssaultDrop() {
     @Override
     public boolean isLocationProhibited(Coords c, int currElevation) {
         Hex hex = game.getBoard().getHex(c);
+        if (hex == null) {
+            return false;
+        }
```
another null check guard that was missing

Scoppio commented on megamek/src/megamek/common/MoveStep.java:
```diff
@@ -3323,7 +3323,9 @@ public boolean isMovementPossible(Game game, Coords src, int srcEl, CachedEntity
         final Hex destHex = game.getBoard().getHex(dest);
         final Entity entity = getEntity();
-
+        if (destHex == null) {
+            return false;
+        }
```
I really want to give this program a pass with SonarQube

Scoppio commented on megamek/src/megamek/common/SpecialHexDisplay.java:
```diff
@@ -23,14 +24,16 @@
 
     public enum Type {
-        ARTILLERY_AUTOHIT(new MegaMekFile(Configuration.hexesDir(), "artyauto.gif").toString()) {
+        ARTILLERY_AUTOHIT(new MegaMekFile(Configuration.hexesDir(), "artyauto.gif")) {
```
so far, changing this from string to MegaMegFile seem to be working as expected. This also gave more flexibility on how to access those image files.

Scoppio commented on megamek/src/megamek/common/SpecialHexDisplay.java:
```diff
@@ -220,6 +267,13 @@ public Type getType() {
         return type;
     }
 
+    public Image getDefaultImage() {
+        if (type.useFolderStructure()) {
+            return type.getImage(imageSignature);
+        }
+        return type.getDefaultImage();
+    }
```
allows to access files in folders, very handy for images that are split in multiple pieces

repligator commented:
"You cannot start a fire in a tornado" gets printed to the round report movement phase.

Scoppio commented:
> "You cannot start a fire in a tornado" gets printed to the round report movement phase.

This is interesting and new... the only way for it to print this message is if it tries to ignite a hex, let me run a test here quickly.

repligator commented:
Not-sure-if-intended, but using the commands can cause atmospheric weather in space.
![Screenshot_20241117_165336](https://github.com/user-attachments/assets/27fb4c3a-edf8-4d51-83db-c9038baa6958)

Scoppio commented:
> Not-sure-if-intended, but using the commands can cause atmospheric weather in space. ![Screenshot_20241117_165336](https://private-user-images.githubusercontent.com/114337591/387009144-27fb4c3a-edf8-4d51-83db-c9038baa6958.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzE4ODE2MTksIm5iZiI6MTczMTg4MTMxOSwicGF0aCI6Ii8xMTQzMzc1OTEvMzg3MDA5MTQ0LTI3ZmI0YzNhLWVkZjgtNGQ1MS04M2RiLWM5MDM4YmFhNjk1OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTE3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTExN1QyMjA4MzlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yN2M4N2FhNDRjNDk3MzNjNGM3YTUzOTQzODcwZjYzNDAwYTFjYTNlNjI4MWRhNWY4M2NkY2M1NjRjZjQzN2E0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.bxxJkJAEKOACnVZpFmkZChCc7zoTNddF6mX3cZRHb1Q)

If the server currently allows you to set planetary conditions in space... I'd say its either a feature or a bug, could go both ways.

About the cant start fire in a tornado, it seems to be correct IF the current weather is a tornado. It will print in the phase you are, so if you run the command to start a fire or to start a firestorm while a tornado is happening then the server won't allow the hexes to ignite

repligator commented:
Using an orbital bombardment with a larger than default radius will throw _a lot_ of errors in the log, and won't render correctly.

`17:03:57,152 ERROR [megamek.common.util.ImageUtil] {AWT-EventQueue-0}
megamek.common.util.ImageUtil$AWTImageLoader.loadImage(ImageUtil.java:289) - Trying to load image for a non-existent file data/images/hexes/orbital_bombardment/hit/col_9_row_8_odd.png`

![Screenshot_20241117_171008](https://github.com/user-attachments/assets/712d9edf-5913-4182-980d-1986ae42972d)

Scoppio commented:
Fixed error with radius different than 4, it uses the bomb-hit graphic for other radii values (ortillery with default radius keeps the original image)

Added safeguards for the firestorm, firestart and change weather, now they cant be activated when in space (they still work for ground and atmosphere games).

HammerGS reviewed (approved):
Tested and everything seems to work.

HammerGS reviewed (approved):
Tested and everything seems to work.
